File: .pub-cache/hosted/pub.dev/zego_express_engine-3.23.0/android/src/main/java/im/zego/zego_express_engine/ZegoAudioMixingDataManager.java
package im.zego.zego_express_engine;

import im.zego.zegoexpress.callback.IZegoAudioMixingHandler;
import im.zego.zegoexpress.constants.ZegoAudioChannel;
import im.zego.zegoexpress.constants.ZegoAudioSampleRate;
import im.zego.zegoexpress.entity.ZegoAudioMixingData;
public class ZegoAudioMixingDataManager extends IZegoAudioMixingHandler {

    private static ZegoAudioMixingDataManager instance;
    private IZegoFlutterAudioMixingHandler handler;
    private ZegoAudioMixingData audioMixingData;

    private ZegoAudioMixingDataManager() {

    }

    public static ZegoAudioMixingDataManager getInstance() {
        if (instance == null) {
            instance = new ZegoAudioMixingDataManager();
        }
        return instance;
    }

    public void setAudioMixingHandler(IZegoFlutterAudioMixingHandler handler) {
        if (handler == null) {
            return;
        }
        this.handler = handler;
    }

    @Override
    public ZegoAudioMixingData onAudioMixingCopyData(int expectedDataLength) {
        if (handler == null) {
            return null;
        }
        ZGFlutterAudioMixData data = handler.onAudioMixingCopyData(expectedDataLength);
        if (data == null) {
            return null;
        }
        if (audioMixingData == null) {
            audioMixingData = new ZegoAudioMixingData();
        }
        audioMixingData.audioData = data.audioData;
        audioMixingData.audioDataLength = data.audioDataLength;
        audioMixingData.SEIData = data.SEIData;
        audioMixingData.SEIDataLength = data.SEIDataLength;
        audioMixingData.param.sampleRate = ZegoAudioSampleRate.getZegoAudioSampleRate(data.param.sampleRate.value());
        audioMixingData.param.channel = ZegoAudioChannel.getZegoAudioChannel(data.param.channel.value());
        return audioMixingData;
    }
}

File: .pub-cache/hosted/pub.dev/zego_express_engine-3.23.0/windows/ZegoAudioMixingDataManager.cpp
#include "include/zego_express_engine/ZegoAudioMixingDataManager.h"
#include "internal/ZegoLog.h"
#include <ZegoExpressSDK.h>

static std::shared_ptr<ZegoAudioMixingDataManager> instance_ = nullptr;
static std::once_flag singletonFlag;

class ZegoAudioMixingHandler : public ZEGO::EXPRESS::IZegoAudioMixingHandler {
public:
    void setHandler(std::shared_ptr<IZegoFlutterAudioMixingHandler> handler) {
        handler_ = handler;
    }

private:
    std::shared_ptr<IZegoFlutterAudioMixingHandler> handler_ = nullptr;

    void onAudioMixingCopyData(ZEGO::EXPRESS::ZegoAudioMixingData *data) override {
        if (handler_ == nullptr) {
            return;
        }

        ZGFlutterAudioMixingData result;
        result.audioData = data->audioData;
        result.audioDataLength = data->audioDataLength;
        result.SEIData = data->SEIData;
        result.SEIDataLength = data->SEIDataLength;
        result.param.sampleRate = (ZGFlutterAudioSampleRate)data->param.sampleRate;
        result.param.channel = (ZGFlutterAudioChannel)data->param.channel;

        handler_->onAudioMixingCopyData(&result);

        data->audioDataLength = result.audioDataLength;
        data->param.sampleRate = ZEGO::EXPRESS::ZegoAudioSampleRate(result.param.sampleRate);
        data->param.channel = ZEGO::EXPRESS::ZegoAudioChannel(result.param.channel);
        data->SEIDataLength = result.SEIDataLength;
    }
};

std::shared_ptr<ZegoAudioMixingDataManager> ZegoAudioMixingDataManager::getInstance() {
    std::call_once(singletonFlag, [&] {
        instance_ = std::make_shared<ZegoAudioMixingDataManager>();
    });
    return instance_;
}

std::shared_ptr<ZEGO::EXPRESS::IZegoAudioMixingHandler> ZegoAudioMixingDataManager::getHandler() {
    return zegoHandler_;
}

void ZegoAudioMixingDataManager::setAudioMixingHandler(std::shared_ptr<IZegoFlutterAudioMixingHandler> handler) {
    ZF::logInfo("[AudioMixingDataManager][setAudioMixingHandler] handler: %p", handler.get());
    handler_ = handler;
    if (handler_ && !zegoHandler_) {
        zegoHandler_ = std::make_shared<ZegoAudioMixingHandler>();
    }
    dynamic_cast<ZegoAudioMixingHandler*>(zegoHandler_.get())->setHandler(handler_);
}